File: app/src/main/java/net/veldor/flibustaloader/ui/fragments/SubscriptionsFragment.java
package net.veldor.flibustaloader.ui.fragments;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.RadioGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.SwitchCompat;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import net.veldor.flibustaloader.App;
import net.veldor.flibustaloader.R;
import net.veldor.flibustaloader.adapters.SubscribesAdapter;
import net.veldor.flibustaloader.selections.SubscriptionItem;
import net.veldor.flibustaloader.utils.MyPreferences;
import net.veldor.flibustaloader.utils.SubscribeAuthors;
import net.veldor.flibustaloader.utils.SubscribeBooks;
import net.veldor.flibustaloader.utils.SubscribeSequences;
import net.veldor.flibustaloader.view_models.SubscriptionsViewModel;

import java.util.ArrayList;

public class SubscriptionsFragment extends Fragment {

    private View mRoot;
    private EditText mSubscribeInput;
    private SubscriptionsViewModel mViewModel;
    private RadioGroup mRadioContainer;
    private RecyclerView mRecycler;


    private SubscribeBooks mBooksSubscribeContainer;
    private SubscribesAdapter mSubscribesAdapter;
    private SubscribeAuthors mAuthorsSubscribeContainer;
    private SubscribeSequences mSequencesSubscribeContainer;

    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container, Bundle savedInstanceState) {

        mViewModel = new ViewModelProvider(this).get(SubscriptionsViewModel.class);

        mRoot = inflater.inflate(R.layout.fragment_subscribe, container, false);

        if(mRoot != null){
            Log.d("surprise", "SubscriptionsFragment onCreate 32: setup ui");
            setupUI();

            FragmentActivity activity = getActivity();

            if(activity != null){
                // буду отслеживать изменения списка книг
                LiveData<Boolean> refresh = mBooksSubscribeContainer.mListRefreshed;
                refresh.observe(getActivity(), aBoolean -> {
                    if(aBoolean != null && aBoolean && mRadioContainer.getCheckedRadioButtonId() == R.id.searchBook){
                        refreshSubscriptionList();
                    }
                });

                // буду отслеживать изменения списка авторов
                LiveData<Boolean> authorRefresh = mAuthorsSubscribeContainer.mListRefreshed;
                authorRefresh.observe(getActivity(), aBoolean -> {
                    if(aBoolean != null && aBoolean && mRadioContainer.getCheckedRadioButtonId() == R.id.searchAuthor){
                        refreshAuthorSubscriptionList();
                    }
                });

                // буду отслеживать изменения списка серий
                LiveData<Boolean> sequenceRefresh = mSequencesSubscribeContainer.mListRefreshed;
                sequenceRefresh.observe(getActivity(), aBoolean -> {
                    if(aBoolean != null && aBoolean && mRadioContainer.getCheckedRadioButtonId() == R.id.searchSequence){
                        refreshSequenceSubscriptionList();
                    }
                });

            }
        }
        return mRoot;
    }

    private void refreshAuthorSubscriptionList() {
        ArrayList<SubscriptionItem> autocompleteValues = mAuthorsSubscribeContainer.getSubscribes();
        mSubscribesAdapter.changeList(autocompleteValues);
        mSubscribesAdapter.notifyDataSetChanged();
    }

    private void refreshSequenceSubscriptionList() {
        ArrayList<SubscriptionItem> autocompleteValues = mSequencesSubscribeContainer.getSubscribes();
        mSubscribesAdapter.changeList(autocompleteValues);
        mSubscribesAdapter.notifyDataSetChanged();
    }

    private void refreshSubscriptionList() {
        ArrayList<SubscriptionItem> autocompleteValues = mBooksSubscribeContainer.getSubscribes();
        mSubscribesAdapter.changeList(autocompleteValues);
        mSubscribesAdapter.notifyDataSetChanged();
    }

    private void setupUI() {

        mRecycler = mRoot.findViewById(R.id.resultsList);
        mRecycler.setLayoutManager(new LinearLayoutManager(getContext()));

        mBooksSubscribeContainer = App.getInstance().getBooksSubscribe();
        mAuthorsSubscribeContainer = App.getInstance().getAuthorsSubscribe();
        mSequencesSubscribeContainer = App.getInstance().getSequencesSubscribe();
        showBooks();

        // отслежу переключение типа подписки
        mRadioContainer = mRoot.findViewById(R.id.subscribe_type);
        if(mRadioContainer != null){
            mRadioContainer.setOnCheckedChangeListener((group, checkedId) -> {
                if (checkedId == R.id.searchBook) {
                    showBooks();
                } else if (checkedId == R.id.searchAuthor) {
                    showAuthors();
                } else if (checkedId == R.id.searchSequence) {
                    showSequences();
                }
            });
        }

        // обработаю добавление книги в список загрузки
        // добавлю идентификатор строки поиска
        mSubscribeInput = mRoot.findViewById(R.id.subscribe_name);

        Button subscribeBtn = mRoot.findViewById(R.id.add_to_blacklist_btn);
        if(subscribeBtn != null){
            subscribeBtn.setOnClickListener(this::addSubscribe);
        }

        // назначу действие переключателю автоматической подписки
        SwitchCompat switcher = mRoot.findViewById(R.id.switchAutoCheckSubscribes);
        if(switcher != null){
            switcher.setChecked(MyPreferences.getInstance().isSubscriptionsAutoCheck());
            switcher.setOnCheckedChangeListener((buttonView, isChecked) -> {
                MyPreferences.getInstance().switchSubscriptionsAutoCheck();
                mViewModel.switchSubscriptionsAutoCheck();
                if(isChecked){
                    Toast.makeText(getContext(), "Подписки будут автоматически проверяться раз в сутки",Toast.LENGTH_SHORT).show();
                }
                else{
                    Toast.makeText(getContext(), "Автоматическая провека подписок отключена, чтобы проверить- нажмите на кнопки выше",Toast.LENGTH_SHORT).show();
                }
            });
        }
        // назначу действие кнопкам проверки подписок
        Button fullSubCheckBtn = mRoot.findViewById(R.id.totalCheckButton);
        if(fullSubCheckBtn != null){
            fullSubCheckBtn.setOnClickListener(v -> {
                Toast.makeText(getContext(), "Выполняю поиск подписок по всем новинкам, это займёт время. Найденные результаты будут отображаться во вкладке 'Найденное'.",Toast.LENGTH_SHORT).show();
                v.setEnabled(false);
                mViewModel.fullCheckSubscribes();
            });
        }
        // назначу действие кнопкам проверки подписок
        Button subCheckBtn = mRoot.findViewById(R.id.fastCheckButton);
        if(fullSubCheckBtn != null){
            subCheckBtn.setOnClickListener(v -> {
                Toast.makeText(getContext(), "Выполняю поиск по книгам, которые поступили после последней проверки. Найденные результаты будут отображаться во вкладке 'Найденное'.",Toast.LENGTH_SHORT).show();
                v.setEnabled(false);
                mViewModel.checkSubscribes();
            });
        }
    }

    private void showBooks() {
        // получу подписки на книги
        ArrayList<SubscriptionItem> autocompleteValues = mBooksSubscribeContainer.getSubscribes();
        mSubscribesAdapter = new SubscribesAdapter(autocompleteValues);
        mRecycler.setAdapter(mSubscribesAdapter);
    }

    private void showAuthors() {
        // получу подписки на авторов
        ArrayList<SubscriptionItem> autocompleteValues = mAuthorsSubscribeContainer.getSubscribes();
        mSubscribesAdapter = new SubscribesAdapter(autocompleteValues);
        mRecycler.setAdapter(mSubscribesAdapter);
    }

    private void showSequences() {
        // получу подписки на серии
        ArrayList<SubscriptionItem> autocompleteValues = mSequencesSubscribeContainer.getSubscribes();
        mSubscribesAdapter = new SubscribesAdapter(autocompleteValues);
        mRecycler.setAdapter(mSubscribesAdapter);
    }

    public void addSubscribe(View view) {
        String value = mSubscribeInput.getText().toString().trim();
        if(!value.isEmpty()){
            // добавлю подписку в зависимости от типа
            int checkedRadioButtonId = mRadioContainer.getCheckedRadioButtonId();
            if (checkedRadioButtonId == R.id.searchBook) {
                Log.d("surprise", "SubscribeActivity addSubscribe add book");
                mBooksSubscribeContainer.addValue(value);
            } else if (checkedRadioButtonId == R.id.searchAuthor) {
                Log.d("surprise", "SubscribeActivity addSubscribe add author");
                mAuthorsSubscribeContainer.addValue(value);
            } else if (checkedRadioButtonId == R.id.searchSequence) {
                Log.d("surprise", "SubscribeActivity addSubscribe add sequence");
                mSequencesSubscribeContainer.addValue(value);
            }
            mSubscribeInput.setText("");
            Toast.makeText(getContext(), "Добавляю значение " + value, Toast.LENGTH_LONG).show();
        }
        else{
            Toast.makeText(getContext(), "Введите значение для подписки", Toast.LENGTH_LONG).show();
            mSubscribeInput.requestFocus();
        }
    }
}

File: app/src/main/java/net/veldor/flibustaloader/ui/fragments/SubscriptionsFragment.kt
package net.veldor.flibustaloader.ui.fragments

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.*
import android.view.inputmethod.EditorInfo
import android.widget.*
import androidx.core.content.res.ResourcesCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import co.mobiwise.materialintro.shape.Focus
import co.mobiwise.materialintro.shape.FocusGravity
import co.mobiwise.materialintro.shape.ShapeType
import co.mobiwise.materialintro.view.MaterialIntroView
import net.veldor.flibustaloader.R
import net.veldor.flibustaloader.adapters.SubscribesAdapter
import net.veldor.flibustaloader.databinding.FragmentSubscribeBinding
import net.veldor.flibustaloader.utils.*
import net.veldor.flibustaloader.view_models.SubscriptionsViewModel

class SubscriptionsFragment : Fragment() {
    lateinit var binding: FragmentSubscribeBinding
    private lateinit var mViewModel: SubscriptionsViewModel
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?, savedInstanceState: Bundle?
    ): View {

        mViewModel = ViewModelProvider(this).get(SubscriptionsViewModel::class.java)
        binding = FragmentSubscribeBinding.inflate(inflater, container, false)
        val root = binding.root

        setupUI()
        setupObservers()


        return root
    }

    private fun setupObservers() {
        // буду отслеживать изменения списка книг
        SubscribeBooks.instance.liveSubscribeListAdd.observe(viewLifecycleOwner, {
            if (it != null) {
                if (binding.subscribeType.checkedRadioButtonId == R.id.searchBook) {
                    (binding.resultsList.adapter as SubscribesAdapter).itemAdded(it)
                    SubscribeBooks.instance.liveSubscribeListAdd.value = null
                }
            }
        })
        SubscribeBooks.instance.liveSubscribeListRemove.observe(viewLifecycleOwner, {
            if (binding.subscribeType.checkedRadioButtonId == R.id.searchBook) {
                (binding.resultsList.adapter as SubscribesAdapter).itemRemoved(it)
            }
        })
        // буду отслеживать изменения списка авторов
        SubscribeAuthors.instance.liveSubscribeListAdd.observe(viewLifecycleOwner, {
            if (it != null) {
                if (binding.subscribeType.checkedRadioButtonId == R.id.searchAuthor) {
                    (binding.resultsList.adapter as SubscribesAdapter).itemAdded(it)
                    SubscribeAuthors.instance.liveSubscribeListAdd.value = null
                }
            }
        })
        SubscribeAuthors.instance.liveSubscribeListRemove.observe(viewLifecycleOwner, {
            if (binding.subscribeType.checkedRadioButtonId == R.id.searchAuthor) {
                (binding.resultsList.adapter as SubscribesAdapter).itemRemoved(it)
            }
        })
        // буду отслеживать изменения списка серий
        SubscribeSequences.instance.liveSubscribeListAdd.observe(viewLifecycleOwner, {
            if (it != null) {
                if (binding.subscribeType.checkedRadioButtonId == R.id.searchSequence) {
                    (binding.resultsList.adapter as SubscribesAdapter).itemAdded(it)
                    SubscribeSequences.instance.liveSubscribeListAdd.value = null
                }
            }
        })
        SubscribeSequences.instance.liveSubscribeListRemove.observe(viewLifecycleOwner, {
            if (binding.subscribeType.checkedRadioButtonId == R.id.searchSequence) {
                (binding.resultsList.adapter as SubscribesAdapter).itemRemoved(it)
            }
        })
        // буду отслеживать изменения списка жанров
        SubscribeGenre.instance.liveSubscribeListAdd.observe(viewLifecycleOwner, {
            if (it != null) {
                if (binding.subscribeType.checkedRadioButtonId == R.id.searchGenre) {
                    (binding.resultsList.adapter as SubscribesAdapter).itemAdded(it)
                    SubscribeGenre.instance.liveSubscribeListAdd.value = null
                }
            }
        })
        SubscribeGenre.instance.liveSubscribeListRemove.observe(viewLifecycleOwner, {
            if (binding.subscribeType.checkedRadioButtonId == R.id.searchGenre) {
                (binding.resultsList.adapter as SubscribesAdapter).itemRemoved(it)
            }
        })
    }


    private fun setupUI() {
        showHints()
        if(PreferencesHandler.instance.isEInk){
            binding.searchBook.setTextColor(ResourcesCompat.getColor(resources, R.color.black, null))
            binding.searchAuthor.setTextColor(ResourcesCompat.getColor(resources, R.color.black, null))
            binding.searchSequence.setTextColor(ResourcesCompat.getColor(resources, R.color.black, null))
            binding.searchGenre.setTextColor(ResourcesCompat.getColor(resources, R.color.black, null))
        }
        binding.subscribeItemInput.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                // Do whatever you want here
                addSubscribe()
                true
            } else false
        }

        binding.resultsList.layoutManager = LinearLayoutManager(context)
        binding.resultsList.adapter = SubscribesAdapter(SubscribeBooks.instance.getSubscribes())
        binding.subscribeType.setOnCheckedChangeListener { _: RadioGroup?, checkedId: Int ->
            when (checkedId) {
                R.id.searchBook -> {
                    (binding.resultsList.adapter as SubscribesAdapter).changeList(SubscribeBooks.instance.getSubscribes())
                }
                R.id.searchAuthor -> {
                    (binding.resultsList.adapter as SubscribesAdapter).changeList(SubscribeAuthors.instance.getSubscribes())
                }
                R.id.searchSequence -> {
                    (binding.resultsList.adapter as SubscribesAdapter).changeList(SubscribeSequences.instance.getSubscribes())
                }
                R.id.searchGenre -> {
                    (binding.resultsList.adapter as SubscribesAdapter).changeList(SubscribeGenre.instance.getSubscribes())
                }
            }
        }

        binding.addSubscriptionBtn.setOnClickListener { addSubscribe() }

        // назначу действие переключателю автоматической подписки
        binding.switchAutoCheckSubscribes.isChecked =
            PreferencesHandler.instance.isSubscriptionsAutoCheck
        binding.switchAutoCheckSubscribes.setOnCheckedChangeListener { _: CompoundButton?, isChecked: Boolean ->
            PreferencesHandler.instance.isSubscriptionsAutoCheck =
                !PreferencesHandler.instance.isSubscriptionsAutoCheck
            mViewModel.switchSubscriptionsAutoCheck()
            if (isChecked) {
                Toast.makeText(
                    context,
                    "Подписки будут автоматически проверяться раз в сутки",
                    Toast.LENGTH_SHORT
                ).show()
            } else {
                Toast.makeText(
                    context,
                    "Автоматическая провека подписок отключена, чтобы проверить- нажмите на кнопки выше",
                    Toast.LENGTH_SHORT
                ).show()
            }
        }
    }

    private fun showHints() {
        showFirstHelp()
    }

    private fun addSubscribe() {
        val value = binding.subscribeItemInput.text.toString().trim { it <= ' ' }
        if (value.isNotEmpty()) {
            // добавлю подписку в зависимости от типа
            when (binding.subscribeType.checkedRadioButtonId) {
                R.id.searchBook -> {
                    SubscribeBooks.instance.addValue(value)
                }
                R.id.searchAuthor -> {
                    SubscribeAuthors.instance.addValue(value)
                }
                R.id.searchSequence -> {
                    SubscribeSequences.instance.addValue(value)
                }
                R.id.searchGenre -> {
                    SubscribeGenre.instance.addValue(value)
                }
            }
            binding.subscribeItemInput.setText("")
            Toast.makeText(context, "Добавляю значение $value", Toast.LENGTH_LONG).show()
        } else {
            Toast.makeText(context, "Введите значение для подписки", Toast.LENGTH_LONG).show()
            binding.subscribeItemInput.requestFocus()
        }
    }
}


private fun SubscriptionsFragment.showFirstHelp() {
    Handler(Looper.getMainLooper()).postDelayed({
        kotlin.run {
            MaterialIntroView.Builder(requireActivity())
                .enableDotAnimation(true)
                .enableIcon(false)
                .setFocusGravity(FocusGravity.CENTER)
                .setFocusType(Focus.MINIMUM)
                .setDelayMillis(300)
                .setUsageId("subscribe type select")
                .enableFadeAnimation(true)
                .performClick(true)
                .setListener {
                    showSecondHelp()
                }
                .setInfoText(getString(R.string.select_subscription_type_message))
                .setTarget(binding.searchBook)
                .setShape(ShapeType.CIRCLE)
                .show()
        }
    }, 100)
}

private fun SubscriptionsFragment.showSecondHelp() {
    Handler(Looper.getMainLooper()).postDelayed({
        kotlin.run {
            MaterialIntroView.Builder(requireActivity())
                .enableDotAnimation(true)
                .enableIcon(false)
                .setFocusGravity(FocusGravity.CENTER)
                .setFocusType(Focus.MINIMUM)
                .setDelayMillis(700)
                .setUsageId("subscribe value enter")
                .enableFadeAnimation(true)
                .performClick(true)
                .setListener {

                }
                .setInfoText(getString(R.string.subscription_second_hint_text))
                .setTarget(binding.subscribeItemInput)
                .setShape(ShapeType.CIRCLE)
                .show()
        }
    }, 100)
}